docs/specs/bot-protection.md

```diff
@@ -0,0 +1,450 @@
+  fail_open: true
+  provider:
+    type: cloudflare
+  risk_assessment:
+    enabled: true
+    provider:
+      type: recaptchav3
+```
+
+- `bot_protection.mode`: When bot protection is required. See [Risk level `mode`](#risk-level-mode).
+- `bot_protection.fail_open`: If it is true, then if the challenge-based provider is service unavailable, access is granted. It is false by default.
+- `bot_protection.provider.type`: If `mode` is not `never`, then it is required. Specify the challenge-based provider to be used in this branch.
+- `bot_protection.risk_assessment.enabled`: Whether risk assessment is enabled.
+- `bot_protection.risk_assessment.provider.type`: It `enabled` is true, then it is required. Specify the risk assessment provider to be used in this branch.
+
+For example,
+
+```yaml
+authentication_flow:
+  login_flows:
+  - name: default
+    steps:
+    - type: identify
+      one_of:
+      - identification: email
+        # Identify with email requires bot protection.
+        bot_protection:
+          mode: "always"
+          provider:
+            type: cloudflare
+    - type: authenticate
+      one_of:
+      - authentication: primary_password
+      - authentication: primary_oob_otp_email
+```
+
+### Behavior of generated flows
+
+Given `bot_protection.risk_assessment.enabled=true`,
+
+1. All the branches of the first step (that is, the `identify` step, or the `authenticate` step in reauth flow) has `bot_protection.risk_assessment.enabled=true`.
+2. The configured provider is used as `bot_protection.risk_assessment.provider.type`
+
+The bot protection behavior in generated flows depend on
+
+- `bot_protection.enabled`
+- `bot_protection.authentication_flow`
+- `bot_protection.authenticator`
+
+Given `bot_protection.enabled=true`,
+
+|Configuration|Behavior|
+|---|---|
+|`bot_protection.authentication_flow.signup_or_login.mode`|Contribute to branches `identification: email`, `identification: phone`, `identification: username` of step `type: identify` in flows `signup`, `login`, `signup_login`, `promote`.|
+|`bot_protection.authentication_flow.account_recovery.mode`|Contribute to all branches of `type: identify` of flow `account_recovery` has `bot_protection.mode` equal to `bot_protection.authentication_flow.account_recovery.mode`.|
+|`bot_protection.authenticator.password.mode`|Contribute to branches `authentication: primary_password`, `authentication: secondary_password` in step `type: authenticate` in flows `login`.|
+|`bot_protection.authenticator.oob_otp_email.mode`|Contribute to branches `authentication: primary_oob_otp_email`, `authentication: secondary_oob_otp_email` in steps `type: authenticate`, `type: create_authenticator` in flows `signup`, `login`, `promote`.|
+|`verification.claims.email.required`|If `verification.claims.email.required` is true, contribute to branches `identification: email` in step `type: identify` in flows `signup`, `promote`.|
```

Comment:
On 2nd thoughts, 

is there case where
1. `signup identify by email`, but `email` is NOT sent in next immediate step?
2. `signup identify by phone`, but `phone code` is NOT sent in next immediate step?

I tried both `w/` & `w/o` verify, seems an email/phone code are always sent in next immediate step, w/o extra user interaction

Therefore I suggest:

for `bot_protection.authentication_flow.authenticator.oob_otp_email.mode`  (& sms), it should contribute to

1. ALWAYS `identification: email` in step `type: identify` in flows `signup`, `promote`
2. `authentication: primary_oob_otp_email`, `authentication: secondary_oob_otp_email` in steps `type: authenticate` in flows `login`
3. `authentication: secondary_oob_otp_email` in steps `type: create_authenticator` in flows `signup`, `promote`

Note how `primary_oob_otp_email` is ignored in 3, because it is handled by 1 already

Similarly for `sms`

Reply:
> signup identify by email, but email is NOT sent in next immediate step?

Yes, Use password to sign up, and turn off verification.

Reply:
I tried with below custom flow, to no avail.

Error is `invalid value` in authUI enter password page

```yaml
authentication_flow:
  signup_flows:
  - name: default
    steps:
    - name: signup_identify
      type: identify
      one_of:
      - identification: email
        steps:
        - name: authenticate_primary_password
          type: create_authenticator
          one_of:
          - authentication: primary_password
        - name: authenticate_primary_oob_otp_email
          type: create_authenticator
          one_of:
          - authentication: primary_oob_otp_email
```

### base64-decoded error
```
{
  "Form": {
    "gorilla.csrf.Token": [
      "FqOqdERU8uGw1l5ZBY/7syLlIxbDcDVQACY0faDmcIF+y66C3zHKWZwXLxIMXxhto6THBs+pmGQC1hGMtUvX/A=="
    ],
    "q_login_id": ["abc@oursky.com"],
    "q_login_id_input_type": ["text"],
    "q_login_id_key": ["email"],
    "q_login_id_type": ["email"],
    "x_action": ["login_id"]
  },
  "Error": {
    "name": "Invalid",
    "reason": "ValidationFailed",
    "message": "invalid value",
    "code": 400,
    "info": {
      "causes": [
        {
          "location": "",
          "kind": "required",
          "details": {
            "actual": ["authentication", "channel"],
            "expected": ["authentication", "target"],
            "missing": ["target"]
          }
        }
      ]
    }
  }
```